### ISSUE TITLE
RTL and styled-systems

### ISSUE THREAD

**Issue by ShadOoW** [styled-system/styled-system#704]
Hello,

I've recently started using styled-system and I think it is amazing, without it I will not have started using CSS in JS.

I've came across a use case where I want to transform some css for RTL (cssjanus), it works great with styled-components syntax, but I was wondering if there is a way to do it in styled-system.

Current implementation: https://codesandbox.io/s/styled-components-direction-cnt7d

If it is not possible, do you think that this is something that could be added to styled-system?

Sorry for the generic question, any feedback would be appreciated.

**Comment by jxnblk** [comment]
Are you not able to use `direction: rtl` in your app? https://developer.mozilla.org/en-US/docs/Web/CSS/direction

**Comment by ShadOoW** [comment]
In the example I shared, you can see that these props need to be reversed when I change the direction of the page:
```
border-left: 5px dotted red;
margin-right: 50px;
```

to became
```
border-right: 5px dotted red;
margin-left: 50px;
```

Changing dir doesn't fix css props like margin / padding / right / left / border and so on, I have to manually add that css, in the example I'm using a library to automatically generate the reverse css (just to not have to manually type it).

I actually have no idea how styled-system could help with that, so i don't really have a suggestion, I just want to know if you know something about this.

I'm thinking of something like this:
```
<Component
    ltrRight={0} // If a prop or a class is set to rtl then this prop is applied
    rtlLeft={0} // If a prop or a class is set to ltr then this prop is applied
/>
```

In CSS it will translate to something like
```
[dir="ltr"] .iaJlSC {
    right: 0px;
}
[dir="rtl"] .iaJlSC {
    left: 0px;
}
```
The more I think about it the less it makes sense :p
If you can't think of something, just let me know and I will close the issue.

**Comment by ShadOoW** [comment]
In the link you shared direction: rtl and dir="rtl" are the same.

Using dir="rtl" changes the direction of the page, but some css properties are not affected by this.
for example: right / left / padding / margin / border-* and so on.

To truly change the direction of a page you have to also reverse some css.

In the example above I'm using a library to generate the reverse css, just to avoid typing it manually.

I'm thinking of something like this:
```
<Component
   position="absolute"
    ltrRight={0} // Will apply this if page has dir="ltr"
    rtlLeft={0} // Will apply this if page has dir="rtl"
/>
```

Which would generate CSS that looks like this:
```
[dir="ltr"] .iaJlSC {
    right: 0px;
}
[dir="rtl"] .iaJlSC {
    left: 0px;
}
```

I don't see how styled-system can help with this, that why I'm asking you, if you also think that there is nothing that can be done, then I will close the issue.

**Comment by jxnblk** [comment]
It'd be nice if there was a utility built-in for this or if you could use the `system` prop to create something like this. Personally, I'd probably build a custom one from scratch for your example above, e.g.
```js
const rltRight = (props) => {
  return {
    '[dir="rtl"] &': {
      right: props.rtlRight,
    }
  }
}
```

**Comment by ShadOoW** [comment]
hi @jxnblk 

I like the idea of a utility or custom prop, but I can't figure out how to do it by looking at the documentation.
https://styled-system.com/custom-props
The transform function only transforms the value, but I need to add a selector.

```
const rltRight = (props) => {
  return {
    '[dir="rtl"] &': {
      right: props.rtlRight,
    }
  }
}
```

This looks like a function that accept props and return an object, I have no idea where it is suppose to be used in relation to styled-system.

I'm sorry but can you give me more hints, as I can't figure out how to generate custom css classes based on a custom prop :(

**Comment by jxnblk** [comment]
Oh, sorry, just typed that out quickly without the full example. You're right in that the current `system` utility wouldn't be able to handle this, but I think it would be a *nice-to-have* addition to the library.

If you take the example `rtlRight` function from above (sorry about the typo), you can pass that into the `styled` HOC from either `styled-components` or `@emotion/styled`.

```js
const Box = styled('div')(
  rtlRight
)
```

The example I listed does *not* use Styled System – it's a custom function. Let me know if that doesn't make sense.

**Comment by ShadOoW** [comment]
**Awesome thank you so much for your help.**

I was able to implement a decent solution for styled-system, based on the info you gave me, I also created an example for future generation of googlers :D

https://codesandbox.io/s/styled-components-direction-gyq2k

This is by no mean a complete solution, but it good enough for what I need.
I'm still open to suggestion on how to improve it.

I will close the issue.
